031-tls-encrypt-the-kafka-connect-rest-api.md

```diff
@@ -0,0 +1,172 @@
+certificate is not signed by a CA root certificate that is trusted by the Java runtime by default (i.e. signed by a well known CA like
+Verisign or Let's Encrypt).
+
+For example, the following CR will enable the HTTP listener on port 8083 and the HTTPS listener on port 8443:
+```
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaConnect
+spec:
+  restListeners:
+  - tls: true
+    certChainAndKey:
+      secretName: my-secret
+      certificateChain: my-certificate.crt
+      key: my-key.key
+    trustedCertificates:
+      secretName: my-secret
+      certificate: ca.crt
+  - tls: false
+  ...
+```
+
+The schema for `restListeners` will be as follows:
+```
+openAPIV3Schema:
+  type: object
+  properties:
+    spec:
+      type: object
+      properties:
+        restListeners:
+          type: array
+          items:
+            type: object
+            properties:
+              tls:
+                type: boolean
+                description: Enables TLS encryption on the listener.
+              certChainAndKey:
+                type: object
+                properties:
+                  certificateChain:
+                    type: string
+                    description: The name of the certificate file in the Secret.
+                  key:
+                    type: string
+                    description: The name of the private key file in the Secret.
+                  secretName:
+                    type: string
+                    description: The name of the Secret containing the certificate.
+                  required:
+                    - key
+                    - certificateChain
+                    - secretName
+                description: Certificate chain and key for Kafka Connect to use to create a keystore for the rest listener.
+              trustedCertificates:
+                type: object
+                properties:
+                  certificate:
+                    type: string
+                    description: The name of the certificate file in the Secret.
+                  secretName:
+                    type: string
+                    description: The name of the Secret containing the certificate.
+                  required:
+                    - certificate
+                    - secretName
+                description: Certificate for Strimzi to trust when making calls to the Connect rest listener.
+            required:
+            - tls
+          description: List of additional REST listeners.
+        ...
+```
+
+the above CR would result in the following properties in the generated connect configuration:
+
+```
+listeners: https://:8443,http://:8083
+rest.advertised.listener: https
+rest.advertised.port: 8443
+listeners.https.ssl.client.auth: none
+listeners.https.ssl.keystore.location: /tmp/kafka/kafka-connect-rest.keystore.p12
+listeners.https.ssl.keystore.password: ***generated password***
+listeners.https.ssl.keystore.type: PKCS12
+```
```

Comment:
* Does connect support using X509 files directly as KAfka does? If yes, maybe we should prefer that.
* With multi-node connect, the nodes talk with each other as well using the REST API. So this will also need to be configured to trust the other nodes. That should be added here to the config I guess as well.

Reply:
@scholzj do you mean using X509 file as the keystore and truststore files, or specifying X509 certificates directly in the config? The docs are unclear but I tried supplying a PEM file for my keystore and truststore and ran into a NoSuchAlgorithmException, so it seems the answer is no, but I'll keep investigating to make sure.

Reply:
> @scholzj do you mean using X509 file as the keystore and truststore files, or specifying X509 certificates directly in the config? The docs are unclear but I tried supplying a PEM file for my keystore and truststore and ran into a NoSuchAlgorithmException, so it seems the answer is no, but I'll keep investigating to make sure.

Yeah, with the first point I meant using the PEM/X509 files directly (i.e. the key and crt files). Kafka supports it, I'm not sure about Connect, but it sounds like it does not. Maybe something @tombentley or someone can look into in Kafka for the longterm